### ISSUE TITLE
ActiveRecord::Attributes causes ActiveModel::Dirty not to notice String#strip! methods

### ISSUE THREAD

**Issue by patrick-gleeson** [rails/rails#40076]
### Steps to reproduce
```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "activerecord", "6.0.3"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :cars, force: true do |t|
    t.string :description
  end

  create_table :vans, force: true do |t|
    t.string :description
  end
end

class Car < ActiveRecord::Base; end

class Van < ActiveRecord::Base
  attribute :description
end


class BugTest < Minitest::Test
  def test_change_detection
    Car.create! description: "Bentley\n"
    car = Car.first
    car.description.strip!

    assert_equal car.description, "Bentley"
    assert car.description_changed? # Passes

    Van.create! description: "VW Camper\n"
    van = Van.first
    van.description.strip!

    assert_equal van.description, "VW Camper"
    assert van.description_changed? # Fails!
  end
end
```
### Expected behavior
In the above example, because I have changed `van.description` from "VW Camper\n" to "VW Camper", then `van.description_changed?` should return `true`.

### Actual behavior
`van.description_changed?` returns `false`.

### System configuration
**Rails version**: 6.0.3

**Ruby version**: 2.5.5

**Comment by composerinteralia** [comment]
Thanks for the clear reproduction script. What is happening here is that you are mutating `van.description` in place by using `String#strip!`. If you want Active Record to be able to track that you have changed the attribute, you would need to use the setter method along with `String#strip`, which does not mutate the object in place:

```diff
- van.description.strip!
+ van.description = van.description.strip
```

**Comment by composerinteralia** [comment]
I'm going to close this issue, since this is the expected behavior of using `String#strip!`. Thanks!

**Comment by patrick-gleeson** [comment]
Hi @composerinteralia, thanks for your response. The only thing I'd point out is that it's a bit weird that in the example I gave, modifying `car.description` with a `strip!` _does_ cause ActiveRecord to pick up the change. So this behaviour _changes_ depending on whether or not you happen to use the Attributes API to describe the attribute, which is quite inconsistent. At the very least, would it not make sense to document this unexpected side effect of using the Attributes API?

**Comment by patrick-gleeson** [comment]
Or to put it another way, if as you say the expected behavior of String#strip! is to bypass ActiveModel::Dirty's change detection, then the behaviour of the Car object in my repro script above is a bug!

**Comment by composerinteralia** [comment]
Ah, reopening since I missed that.

**Comment by yoelblum** [comment]
@patrick-gleeson i agree this doesn't look right

**Comment by composerinteralia** [comment]
Looking closer, it looks like this is a difference between [`ActiveModel::Type::String`](https://github.com/rails/rails/blob/b5dac96b59a582c7968b0c33d6ff39d1cfc336dc/activemodel/lib/active_model/type/string.rb#L27), which returns a new string, and [`ActiveModel::Type::Value`](https://github.com/rails/rails/blob/master/activemodel/lib/active_model/type/value.rb#L128-L130) which returns back the same object.

You can work around this by changing to a string type:

```diff
- attribute :description
+ attribute :description, :string
```

**Comment by jonathanhefner** [comment]
This should be fixed by #39830 on `master`.
